### PR TITLE
introducing mina-automation for component level testing

### PR DESCRIPTION
**Pull request by dkijania**
Added test library which can testing  on component level some auxiliary apps (especially those which are single threaded and can be used as cli). I've added tiny feature which auto-detect if environment in which we run test is CI or local development. mina-automation library which should be a corner stone for developing tests with single node or small test for corner cases test (which scope is to narrow to promote it to integration test).

### CODE REVIEW

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: src/test/mina_automation/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+mina automation
+==============
+
+The `mina_automation` lib is a small utility which helps to automate local app execution
+
+Library can be used in various local test and automation. For example to test if archive and extract blocks apps gives expected archive state. Main feature of an lib is ability to detect location of apps. App can be build locally and resides in _build/default folder, in /usr/local/bin folder.
```
Comment (georgeee) [inline src/test/mina_automation/README.md:6]:
```suggestion
Library can be used in various local test and automation. For example to test if archive and extract blocks apps gives expected archive state. Main feature of an lib is ability to detect location of apps. App can be build locally and resides in _build/default folder, or can be installed globally and put in /usr/local/bin folder.
```

File: src/test/mina_automation/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+It also has automation for docker if we would like to run docker command locally and use some output further in the test.
+
+An example of usage:
```
Comment (georgeee) [inline src/test/mina_automation/README.md:10]: Could you add description of what this example performs exactly?

Also, what are `~blocks` and `~archive_uri`?

File: src/test/mina_automation/archive_blocks.ml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+open Executor
```
Comment (georgeee) [inline src/test/mina_automation/archive_blocks.ml:1]: There are a few modules around. Could you add a short doc-comment to each of them.

For instance for this one I imagine a comment to be like:

```
(**
Module to run archive_blocks utility for the given list of block files and an archive PostgreSQL database.
*)
```

File: src/test/mina_automation/docker.ml [hunk 4]
```diff
@@ -0,0 +1,130 @@
+open Core
+open Integration_test_lib
+open Async
+open Printf
+
+module HttpPath = struct
```
Comment (georgeee) [inline src/test/mina_automation/docker.ml:6]: Could you add a doc-comment for each submodule here?

File: src/test/mina_automation/mina_automation.ml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+(* mina_automation.ml -- test automation modules *)
+
+module Archive_blocks = Archive_blocks
```
Comment (georgeee) [inline src/test/mina_automation/mina_automation.ml:3]: I think this file can be removed
Such a module will be auto-generated by ocaml compiler

File: src/test/mina_automation/executor.ml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    let logger = Logger.create () in
+    match t.context with
+    | AutoDetect -> (
+        match%bind Sys.file_exists (built_name t) with
+        | `Yes ->
+            [%log debug] "running from _build/default folder"
+              ~metadata:[ ("app", `String (built_name t)) ] ;
+            run_from_local t ~args ?env ()
+        | _ -> (
+            match%bind Sys.file_exists t.official_name with
```
Comment (georgeee) [inline src/test/mina_automation/executor.ml:59]: I wonder if instead of detecting an executable in "official path", i.e. `/usr/local/bin/whatever`, we could instead check `PATH` environment variable.

I.e. something like:

```suggestion
let paths = Option.value_map ~f:(String.split ~on:':') ~default:[] (Sys.getenv "PATH") in
let exists_at_path prefix = match Sys.file_exists (prefix ^ "/" ^ t.official_name) with |`Yes -> Some prefix | _ -> None in
match%bind Deferred.List.find_map ~f:exists_at_path paths with
```

This would enable running this script e.g. from within nix

File: src/test/mina_automation/executor.ml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            | `Yes ->
+                [%log debug] "running from usr/bin "
```
Comment (georgeee) [inline src/test/mina_automation/executor.ml:61]:
```suggestion
              | Some prefix ->  [%log debug] "running from %s" prefix
```

(to be used together with the suggestion above)

**Review by georgeee — CHANGES_REQUESTED** [review]

**Comment by dkijania** [comment]
!ci-build-me

File: src/test/mina_automation/README.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ - `image : String.t` - docker image in which command will be run. If does not exist locally it will be pulled
+ - `cmd: String.t` - command which will be run in docker
+ - `workdir: String.t` - working directory off command
+ - `volume: String.t` - volume mapping between docker and localhost
+ - `network: String.t` - network which docker will be attached to when executing command. In our example we are using `localhost` which allows to connect to db from within docker
+  
+
+### extract blocks
+
+Extract blocks apps dumps blocks stored in database in form of Extensional blocks.
```
Comment (georgeee) [inline src/test/mina_automation/README.md:76]:
```suggestion
Extract blocks application dumps blocks stored in database in form of Extensional blocks.
```

**Review by georgeee — APPROVED** [review]
